### PR TITLE
add support for native events from EventProxy

### LINKED ISSUE #51: EventProxy does not support getting non-erased event
It would be convenient to be able to get a non-erased event type from EventProxy. It seems wasteful to create an erased event when type information is available.

### PR DESCRIPTION
**Pull request by trws**
Fixes #51

Add support for getting the native event type out of EventProxy.
This is now the default, with Event fixed to allow implicit conversion
from native event types and a new `operator Event()` overload on
EventProxy all tests pass in tree, need to see how this impacts codes.

### CODE REVIEW

**Comment by trws** [comment]
@MrBurmark, take a look at this, possibly give it a try building with something that uses events.  It should allow existing code to continue working for all expected patterns, but if we find something that breaks we can re-work.

**Review by mdavis36 — COMMENTED** [review]
I think this looks good and should do what is expected, but will wait for @MrBurmark to give this a spin. Do we want to add some tests for this?

**Comment by MrBurmark** [comment]
Sorry, I haven't gotten back to doing graph stuff again yet. This looks reasonable, I'll try to get back to this sometime next week.

**Comment by trws** [comment]
@MrBurmark, did you ever get a chance to test this? I imagine we want it for the release but I haven't heard anything.  Will look into the appveyor failure but I'd like to hear if this design solves your issue first.

**Comment by trws** [comment]
I *think* this is straightened out enough that we can at least work with it for now.  If you get a chance take a look @MrBurmark.

File: include/camp/resource.hpp
Comment (MrBurmark) [inline include/camp/resource.hpp:181]: why are these not const&?

**Comment by MrBurmark** [comment]
Looks good to me. I added a couple tests in my fork https://github.com/MrBurmark/camp/tree/trws/issue51 if you think its worth bringing those in.

**Comment by trws** [comment]
Actually yeah, if you would push those in here that would be great @MrBurmark.

File: include/camp/resource.hpp
Comment (trws) [reply include/camp/resource.hpp:181]: I can try to switch them to that, but the event being pulled out is not const, and the conversion methods were not const, so there were some issues there.  It really could use a pass to ensure const-correctness, which I will try to do but I'd love another set of eyes passing over EventProxy and Event for that if you have the time.

**Comment by MrBurmark** [comment]
@trws I can't push changes to the branch because its on your fork.

**Comment by trws** [comment]
You should be able to now, you are now on the camp team with write permissions, and this PR is set to allow maintainers to push.

File: include/camp/resource.hpp
Comment (MrBurmark) [reply include/camp/resource.hpp:181]: I wasn't sure if you could delete the copy constructor, copy assignment operator unless it had the normal signature.

Comment (MrBurmark) [reply include/camp/resource.hpp:181]: I took a look over the Event related methods and didn't see anything out of place.

**Comment by MrBurmark** [comment]
Thanks Tom, I pushed up the tests.

File: include/camp/resource.hpp
Comment (MrBurmark) [inline include/camp/resource.hpp:195]: These conflict if Res is Resource

File: include/camp/resource.hpp
Comment (MrBurmark) [inline include/camp/resource.hpp:199]: SFINAE to allow EventProxy<Resource> to work. @trws what do you think of this fix?

Comment (trws) [reply include/camp/resource.hpp:199]: I like it, may want to factor this out into something more easily identifiable but I love the result.

**Comment by trws** [comment]
I'm going to add a quick tweak here, then I think we'll be good to go.

**Comment by MrBurmark** [comment]
@trws That makes sense, and makes it more clear what you will get when you use an erased resource vs a non-erased resource.

File: include/camp/resource.hpp
Comment (MrBurmark) [reply include/camp/resource.hpp]: Ok, I am still confused about the signature change so I read the cppreference copy constructor page. I thought the copy constructor had to be T(T const&,) but I see it can also be T(T&) so deleting either signature should work to delete the copy constructor. I don't see how the EventProxy copy constructor signature matters much here especially since its deleted.

**Review by MrBurmark — APPROVED** [review]

File: include/camp/resource.hpp
Comment (trws) [reply include/camp/resource.hpp]: That makes sense, I'll try switching it back, and as you say that should actually behave just fine.  If it does it certainly makes more sense that way.

Comment (trws) [reply include/camp/resource.hpp]: Seems to behave just fine, so I reverted it.  Assuming tests pass, I'm planning to merge this version.

**Pull request merged**
